### PR TITLE
gossip: miscellaneous reductions in gossip traffic

### LINKED ISSUE #30126: gossip: propagation of high water stamps scales poorly
Every `gossip.{Request,Response}` contains a `HighWaterStamps` that is a `map[NodeID]int64`. This map contains an entry for every node in the cluster. Each entry in the map is ~10 bytes when sent over the wire. So if you have a 3-node cluster, every gossip request/response is at least 30 bytes in size. If you have a 128-node cluster, every gossip request/response is at least 1280 bytes in size. 

It should be possible to only send a delta of the high water stamps on each request/response, though we'll need to be careful to do this only when the recipient knows to handle a partial stamps map (i.e. after #30106 has merged). Gating this on a cluster setting is awkward, but I think we can tie it to another optimization: we only need to send `gossip.Request.Addr` on the initial message for a gossip stream. If we make that a null-able field, we can avoid sending it and use its absence as indication that both sides support partial high water stamps updates. 

Cc @spencerkimball, @a-robinson

### PR DESCRIPTION
**Pull request by petermattis**
Make gossip.Request.{Addr,ClusterID} nullable and only send them on the
first request over a connection. Remove gossip.Response.Addr which was
unused.

Only gossip the cluster ID once per server. The cluster ID never
expires, so there was no need to be re-gossiping it frequently.

Keep track of which high water stamps have been sent to the remote side
of a connection and only send stamps that are newer than those already
sent. This has a dramatic impact on gossip traffic as a cluster gets
larger as usually only 1 or 2 stamps have changed on each gossip message
which results in reducing an O(n) map size to a constant.

In a local 27-node cluster, these changes reduce gossip traffic by
~60%.

Fixes #30126
See #30088

Release note: None

### CODE REVIEW

**Review by a-robinson — CHANGES_REQUESTED** [review]
Looks like a good change to me. I don't see anything you're missing.

> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-:-LMJomlO0vOdpzWlCaTw:b-s26u5u)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/gossip/infostore.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJlzd632C-N7pIEbF0:-LMJlzd632C-N7pIEbF1:b-mxg5mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/infostore.go#L277)):*
> ```Go
> // the infoStore.
> func (is *infoStore) getHighWaterStamps(sent map[roachpb.NodeID]int64) map[roachpb.NodeID]int64 {
> 	copy := make(map[roachpb.NodeID]int64, len(is.highWaterStamps))
> ```

It doesn't likely matter, but `len(is.highWaterStamps)` is much more capacity than is needed in the common case now.
___
*[pkg/gossip/server.go, line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJmMui1w-ULl-0wGHZ:-LMJmMui1w-ULl-0wGH_:b-jyarzt) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L121)):*
> ```Go
> 			return err
> 		}
> 		if args.ClusterID != nil &&
> ```

Should we require that the initial gossip message includes a non-nil cluster ID (even if it's empty for the initial join situation)? That would likely be indicative of a bug, given how the code is currently written.
___
*[pkg/gossip/server.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJnDNxDDttnToOTpyJ:-LMJnDNxDDttnToOTpyK:b7donbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L242)):*
> ```Go
> 			// has one, we'll run the logic below to decide whether to keep the
> 			// connection to it or to forward it elsewhere.
> 			log.Infof(ctx, "received initial cluster-verification connection from %s", args.Addr)
> ```

If `args.Addr` is always non-nil then this will be ok, but it's fragile because if we ever get a nil addr then this will panic. `%v` is safer.
___
*[pkg/gossip/server.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJn6141Q0GoSA7XQP7:-LMJn6141Q0GoSA7XQP8:b-56das7) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L247)):*
> ```Go
> 
> 			var addr util.UnresolvedAddr
> 			if args.Addr != nil {
> ```

Isn't `args.Addr` required to be non-nil here? I'd rather enforce that than potentially have multiple connections sharing the same `nodeMap` entry for an empty address.

File: pkg/gossip/infostore.go [hunk 1]
```diff
@@ -273,10 +273,15 @@ func (is *infoStore) addInfo(key string, i *Info) error {
 // getHighWaterStamps returns a copy of the high water stamps map of
 // gossip peer info maintained by this infostore. Does not modify
 // the infoStore.
```
Comment (spencerkimball) [inline pkg/gossip/infostore.go]: Expand the comment?

File: pkg/gossip/server.go [hunk 2]
```diff
@@ -157,28 +171,30 @@ func (s *server) Gossip(stream Gossip_GossipServer) error {
...
 				log.Infof(ctx, "returning %d info(s) to node %d: %s",
-					infoCount, args.NodeID, extractKeys(delta))
+					infoCount, conn.args.NodeID, extractKeys(delta))
 			}
 			// Ensure that the high water stamps for the remote client are kept up to
 			// date so that we avoid resending the same gossip infos as infos are
 			// updated locally.
 			for _, i := range delta {
-				ratchetHighWaterStamp(args.HighWaterStamps, i.NodeID, i.OrigStamp)
+				// log.Infof(ctx, "sending(n%d): %s %d", conn.args.NodeID, k, i.Size())
```
Comment (spencerkimball) [inline pkg/gossip/server.go]: Remove this debug log line.

File: pkg/gossip/server.go [hunk 3]
```diff
@@ -157,28 +171,30 @@ func (s *server) Gossip(stream Gossip_GossipServer) error {
...
+				ratchetHighWaterStamp(conn.args.HighWaterStamps, i.NodeID, i.OrigStamp)
 			}
 
 			*reply = Response{
 				NodeID:          s.NodeID.Get(),
-				HighWaterStamps: s.mu.is.getHighWaterStamps(),
+				HighWaterStamps: s.mu.is.getHighWaterStamps(conn.sentHighWaterStamps),
 				Delta:           delta,
 			}
+			// log.Infof(ctx, "response: %d %d", reply.Size(), len(reply.HighWaterStamps))
```
Comment (spencerkimball) [inline pkg/gossip/server.go]: Ditto.

File: pkg/gossip/server.go [hunk 4]
```diff
@@ -315,8 +336,9 @@ func (s *server) gossipReceiver(
 
 		*reply = Response{
 			NodeID:          s.NodeID.Get(),
-			HighWaterStamps: s.mu.is.getHighWaterStamps(),
+			HighWaterStamps: s.mu.is.getHighWaterStamps(conn.sentHighWaterStamps),
 		}
+		log.Infof(ctx, "response: %d %d", reply.Size(), len(reply.HighWaterStamps))
```
Comment (spencerkimball) [inline pkg/gossip/server.go]: I assume you'd like to remove this log line as well. Will be chatty-af.

**Review by spencerkimball — APPROVED** [review]

**Review by bdarnell — COMMENTED** [review]
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/store.go, line 1609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMOV4Kd74algIHV2cGq:-LMOV4Kd74algIHV2cGr:b5xomwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/storage/store.go#L1609)):*
> ```Go
> 	}
> 
> 	var gossipedClusterID bool
> ```

This seems like a clumsy way to make this once-only. We could just do it in Store.Start when we discover that we have a replica of the first range.

I'm slightly nervous about getting rid of any re-gossiping here - if we had a bug that dropped some gossip messages, the problem would no longer be self-correcting. How much did this part of the change save?

**Review by a-robinson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/store.go, line 1609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMOV4Kd74algIHV2cGq:-LMOY8Kq06y55CymUSSl:b-6v3dus) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/storage/store.go#L1609)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This seems like a clumsy way to make this once-only. We could just do it in Store.Start when we discover that we have a replica of the first range.

I'm slightly nervous about getting rid of any re-gossiping here - if we had a bug that dropped some gossip messages, the problem would no longer be self-correcting. How much did this part of the change save? 
</blockquote></details>

I was a little worried about this too, but we really are gossiping it quite frequently ever since https://github.com/cockroachdb/cockroach/commit/42cd019427309b77d5ad895c76c58ba684835105. A nice compromise might be to set up a separate timer for gossiping the cluster ID less frequently than the every few seconds we're doing so today.

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/store.go, line 1609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMOV4Kd74algIHV2cGq:-LMObvigC_RniATir2fh:b-g2hqsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/storage/store.go#L1609)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

I was a little worried about this too, but we really are gossiping it quite frequently ever since https://github.com/cockroachdb/cockroach/commit/42cd019427309b77d5ad895c76c58ba684835105. A nice compromise might be to set up a separate timer for gossiping the cluster ID less frequently than the every few seconds we're doing so today.
</blockquote></details>

Yeah, a separate timer might make sense. We have some other things that are gossiped hourly (although I guess if we're really concerned about gossip messages getting dropped, an hour is an awfully long time to wait).

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/gossip/client.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAU7-r1-167:-LMOgTem6umTYIQuExxx:b-yre6k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/client.go#L167)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Why this change?
</blockquote></details>

`ClusterID` is now a pointer while `g.clusterID.Get()` returns a value.
___
*[pkg/gossip/client.go, line 187 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAU7-r1-187:-LMOf2WdD21Jl3_puVi1:b916loz) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/client.go#L187)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Remove?
</blockquote></details>

Yeah, this is still a WIP PR. I'll remove before merging.
___
*[pkg/gossip/infostore.go, line 275 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAUB-r1-275:-LMOgyg6A9B0Gff7HdZ8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/infostore.go#L275)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Expand the comment?
</blockquote></details>

Done.
___
*[pkg/gossip/infostore.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJlzd632C-N7pIEbF0:-LMOgyxl0fRcuyH4oQXc:bj74m29) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/infostore.go#L277)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It doesn't likely matter, but `len(is.highWaterStamps)` is much more capacity than is needed in the common case now.
</blockquote></details>

I doubt it matters.
___
*[pkg/gossip/server.go, line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJmMui1w-ULl-0wGHZ:-LMOfBh7CubXi0E1IyDu:bkifb01) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L121)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Should we require that the initial gossip message includes a non-nil cluster ID (even if it's empty for the initial join situation)? That would likely be indicative of a bug, given how the code is currently written.
</blockquote></details>

We don't always know the cluster ID on the initial gossip message. The remote might be bootstrapping. Perhaps what I should be doing is sending the cluster ID at least once on a gossip connection. 
___
*[pkg/gossip/server.go, line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAUC-r1-188:-LMOh6_nA20SjQGukNU4:b880qha) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L188)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Remove this debug log line.
</blockquote></details>

Will do before merging.
___
*[pkg/gossip/server.go, line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAUC-r1-197:-LMOh87EDLZwHyF4mcnS:b-t7tf0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L197)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Will do before merging.
___
*[pkg/gossip/server.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJnDNxDDttnToOTpyJ:-LMOfTEh0X3Qs1vxrctX:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L242)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

If `args.Addr` is always non-nil then this will be ok, but it's fragile because if we ever get a nil addr then this will panic. `%v` is safer.
</blockquote></details>

Good point. Done.
___
*[pkg/gossip/server.go, line 247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJn6141Q0GoSA7XQP7:-LMOgMmh091Xtlfx7-S2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L247)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Isn't `args.Addr` required to be non-nil here? I'd rather enforce that than potentially have multiple connections sharing the same `nodeMap` entry for an empty address.
</blockquote></details>

Done.
___
*[pkg/gossip/server.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMIX8V7CFNLKDj2xAUC-r1-341:-LMOh9IR0aPf0btCDkX4:b880qha) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L341)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I assume you'd like to remove this log line as well. Will be chatty-af.
</blockquote></details>

Will do before merging.
___
*[pkg/storage/store.go, line 1609 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMOV4Kd74algIHV2cGq:-LMOeHrpC1raMqOY-8Sg:bj8vbhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/storage/store.go#L1609)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, a separate timer might make sense. We have some other things that are gossiped hourly (although I guess if we're really concerned about gossip messages getting dropped, an hour is an awfully long time to wait). 
</blockquote></details>

This wasn't a significant savings, just something odd that seemed worth cleaning up. I've added a separate timer to gossip the cluster ID every minute.

**Review by a-robinson — COMMENTED** [review]
Looks good pending cleanup and backwards compatibility

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-:-LMP8vdG-I8OxrSAjjet:bh9po0w)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/gossip/server.go, line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/30193#-LMJmMui1w-ULl-0wGHZ:-LMP7lZbEgaPq6OYByR_:bcwf31p) ([raw file](https://github.com/cockroachdb/cockroach/blob/409426fcee6f845e8948d3cdbfe668e5ce37bd94/pkg/gossip/server.go#L121)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

We don't always know the cluster ID on the initial gossip message. The remote might be bootstrapping. Perhaps what I should be doing is sending the cluster ID at least once on a gossip connection. 
</blockquote></details>

Right, but in that case it should be an empty cluster ID based on how you wrote the code in client.go, not a nil cluster ID. A nil cluster ID looks unexpected.

Don't worry about it, though, it'd add just as much risk as it'd remove.

**Comment by petermattis** [comment]
FYI, I'm not planning on getting this into 2.1 and will revisit in a few weeks.

**Pull request closed without merging**